package/MDAnalysis/coordinates/RDKit.py

```diff
@@ -95,10 +134,591 @@ def __init__(self, filename, **kwargs):
+    stores it into an RDKit AtomPDBResidueInfo object.
+
+    Parameters
+    ----------
+
+    attr : str
+        Name of the atom attribute in MDAnalysis in the singular form
+    value : object, np.int or np.float
+        Attribute value as found in the AtomGroup
+    mi : rdkit.Chem.rdchem.AtomPDBResidueInfo
+        MonomerInfo object that will store the relevant atom attributes
+    """
+    if isinstance(value, np.generic):
+        # convert numpy types to python standard types
+        value = value.item()
+    if attr == "names":
+        # RDKit needs the name to be properly formatted for a
+        # PDB file (1 letter elements start at col 14)
+        name = re.findall(r'(\D+|\d+)', value)
+        if len(name) == 2:
+            symbol, number = name
+        else:
+            symbol, number = name[0], ""
+        value = "{:>2.2}{:<2.2}".format(symbol, number)
+    # set attribute value in RDKit MonomerInfo
+    rdattr = RDATTRIBUTES[attr]
+    getattr(mi, "Set%s" % rdattr)(value)
+
+
+def _set_atom_property(atom, attr, value):
+    """Saves any attribute and value into an RDKit atom property"""
+    if isinstance(value, (float, np.float)):
+        atom.SetDoubleProp(attr, float(value))
+    elif isinstance(value, (int, np.int)):
+        atom.SetIntProp(attr, int(value))
+    else:
+        atom.SetProp(attr, value)
+
+
+def _infer_bo_and_charges(mol, terminal_atom_indices=[]):
+    """Infer bond orders and formal charges from a molecule.
+
+    Since most MD topology files don't explicitly retain information on bond
+    orders or charges, it has to be guessed from the topology. This is done by
+    looping other each atom and comparing its expected valence to the current
+    valence to get the Number of Unpaired Electrons (NUE).
+    If an atom has a negative NUE, it needs a positive formal charge (-NUE).
+    If two neighbouring atoms have UEs, the bond between them most
+    likely has to be increased by the value of the smallest NUE.
+    If after this process, an atom still has UEs, it's either a radical
+    (because one of its bonds was cut when creating the AtomGroup) or it needs
+    a negative formal charge of -NUE. Since these radical atoms can be detected
+    when looping over the bonds of the AtomGroup, only atoms that are not part
+    of this "terminal_atoms" list will be assigned a negative formal charge.
+
+    Parameters
+    ----------
+
+    mol : rdkit.Chem.rdchem.RWMol
+        The molecule is modified inplace and must have all hydrogens added
+
+    terminal_atom_indices : list
+        List of terminal atoms indices, i.e. atoms at the edges of a molecule
+    """
+
+    for atom in mol.GetAtoms():
+        # get NUE for each possible valence
+        expected_vs = PERIODIC_TABLE.GetValenceList(atom.GetAtomicNum())
+        current_v = atom.GetTotalValence() - atom.GetFormalCharge()
+        nue = [v - current_v for v in expected_vs]
```

Comment:
This same basic calculation is repeated several times here to get/update NUEs, so it might be worth making it its own function?